### PR TITLE
Do not resolve the interface name globally

### LINKED ISSUE #3191: Exception when calling sniff for multiple interfaces
### Brief description

Basic sniff capture using multiple interfaces.
When passing a list (or dict) of interfaces, Scapy raises an exception (see below).

See 'Related resources' for probable cause/fix.

#### Environment

Scapy version: 2.4.5
Python version: 3.8.5
Operation System: Ubuntu 20.04.2.0 Desktop

#### How to reproduce

`packets = sniff(iface=["wlan0", "wlan1", "wlan2", "wlan3"], prn=callback, store=0)`

#### Actual result

> File "scapy_file.py", line 197, in run
>     packets = sniff(iface=["wlan0","wlan1","wlan2","wlan3"], prn=callback, store=0)
>   File "/root/.local/share/virtualenvs/myenv/lib/python3.8/site-packages/scapy/sendrecv.py", line 1264, in sniff
>     sniffer._run(*args, **kwargs)
>   File "/root/.local/share/virtualenvs/myenv/lib/python3.8/site-packages/scapy/sendrecv.py", line 1128, in _run
>     sniff_sockets[L2socket(type=ETH_P_ALL, iface=iface,
>   File "/root/.local/share/virtualenvs/myenv/lib/python3.8/site-packages/scapy/arch/linux.py", line 501, in __init__
>     set_promisc(self.ins, self.iface)
>   File "/root/.local/share/virtualenvs/myenv/lib/python3.8/site-packages/scapy/arch/linux.py", line 181, in set_promisc
>     mreq = struct.pack("IHH8s", get_if_index(iff), PACKET_MR_PROMISC, 0, b"")
>   File "/root/.local/share/virtualenvs/myenv/lib/python3.8/site-packages/scapy/arch/linux.py", line 401, in get_if_index
>     return int(struct.unpack("I", get_if(iff, SIOCGIFINDEX)[16:20])[0])
>   File "/root/.local/share/virtualenvs/myenv/lib/python3.8/site-packages/scapy/arch/common.py", line 65, in get_if
>     return ioctl(sck, cmd, struct.pack("16s16x", iff.encode("utf8")))
> AttributeError: 'list' object has no attribute 'encode'

#### Expected result

Expecting sniff to begin.

#### Related resources

Source probably related to this issue:
[sendrecv.py - line 1111 to line 1128](https://github.com/secdev/scapy/blob/master/scapy/sendrecv.py#L1111)

Since I don't use a custom L2 socket, I simply commented out line 1111 to 1113 and used the default L2 socket:
`L2socket = conf.L2listen`

Proper fix probably involves looping through list items to instance an L2socket.

### PR DESCRIPTION
**Pull request by guedou**
This PR fixes #3191. I do not know how to test it =/

### CODE REVIEW

**Comment by polybassa** [comment]
If you need to create multiple interfaces to create a unit test, you could use virtual CAN interfaces. At least on Linux machines. 
You can create as many devices as you want:
```
sudo modprobe vcan
sudo ip link add dev vcan0 type vcan
sudo ip link set up vcan0
sudo ip link add dev anotherinterface type vcan
sudo ip link set up anotherinterface
sudo ip link add dev onemoreinterface type vcan
sudo ip link set up onemoreinterface
```

```
sniff(["vcan0", "anotherinterface", "onemoreinterface"])
```

If you think that could work for a testcase, I could provide one.

File: scapy/sendrecv.py [hunk 1]
```diff
@@ -1108,9 +1108,8 @@ def _run(self,
                             quiet=quiet)
                 )] = offline
         if not sniff_sockets or iface is not None:
-            iface = resolve_iface(iface or conf.iface)
             if L2socket is None:
-                L2socket = iface.l2listen()
+                L2socket = resolve_iface(iface or conf.iface).l2listen()
```
Comment (gpotter2) [inline scapy/sendrecv.py:1112]: Don't we need to put this in the `isinstance()`s to make sure it always work?

```
if isinstance(iface, list):
    L2socket = resolve_iface(iface[0] or conf.iface).l2listen()
...
elif isinstance(iface, dict):
    L2socket = resolve_iface(next(iter(iface.values()) or conf.iface).l2listen()
....
.....
```

**Comment by gpotter2** [comment]
We can also just use TUN/TAPS polybassa, but good call.

**Comment by gpotter2** [comment]
I'd go for
```
diff --git a/scapy/sendrecv.py b/scapy/sendrecv.py
index 7a847a03..130f8700 100644
--- a/scapy/sendrecv.py
+++ b/scapy/sendrecv.py
@@ -1108,22 +1108,22 @@ class AsyncSniffer(object):
                             quiet=quiet)
                 )] = offline
         if not sniff_sockets or iface is not None:
-            if L2socket is None:
-                L2socket = resolve_iface(iface or conf.iface).l2listen()
+            _RL2 = lambda i: L2socket or resolve_iface(iface or
+                    conf.iface).l2listen()
             if isinstance(iface, list):
                 sniff_sockets.update(
-                    (L2socket(type=ETH_P_ALL, iface=ifname, **karg),
+                    (_RL2(ifname)(type=ETH_P_ALL, iface=ifname, **karg),
                      ifname)
                     for ifname in iface
                 )
             elif isinstance(iface, dict):
                 sniff_sockets.update(
-                    (L2socket(type=ETH_P_ALL, iface=ifname, **karg),
+                    (_RL2(ifname)(type=ETH_P_ALL, iface=ifname, **karg),
                      iflabel)
                     for ifname, iflabel in six.iteritems(iface)
                 )
             else:
-                sniff_sockets[L2socket(type=ETH_P_ALL, iface=iface,
+                sniff_sockets[_RL2(iface)(type=ETH_P_ALL, iface=iface,
                                        **karg)] = iface

         # Get select information from the sockets
```

**Pull request closed without merging**